### ISSUE TITLE
[SCM-772] GitStatusConsumer does not properly handle quoted paths output from GitStatusCommand

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-scm#979]
**[Gary L Peskin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=garyp@firstech.com)** opened **[SCM-772](https://issues.apache.org/jira/browse/SCM-772?redirect=false)** and commented

We have several files in our repository with spaces in the file name.  Even with --porcelain specified, the filename appears with quotes around it.  Here is an example:

```
M src/main/webapp/admin-index.html
M src/main/webapp/adminhelp/en_US/css/AboutThisBook.css
M "src/main/webapp/adminhelp/en_US/css/Appendix A Role Manager Defs.css"
```

This messes up the regular expression parsing in GitStatusConsumer.

A better approach would be to modify GitStatusCommand to specify the `"\-z"` option instead of the `"--porcelain"` option.  This option is specifically designed to generate machine parseable output from the git status command.  This would necessitate changes to GitStatusConsumer but would allow the elimination of all of the regex parsing in GitStatusConsumer making for a more reliable solution.

If you'd like, please let me know and I can work up a patch for the changes to these two classes and send it on.

---

**Affects:** 1.9, 1.9.1

**Attachments:**
- [maven-scm.patch](https://issues.apache.org/jira/secure/attachment/12718911/maven-scm.patch) (_41.32 kB_)

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

From http://git-scm.com/docs/git-status:

> ## Porcelain Format
> 
> The porcelain format is similar to the short format, but is guaranteed not to change in a backwards-incompatible way between Git versions or based on user configuration. This makes it ideal for parsing by scripts. The description of the short format above also describes the porcelain format, with a few exceptions:
> 
> 1.The user's color.status configuration is not respected; color will always be off.
> 2.The user's status.relativePaths configuration is not respected; paths shown will always be relative to the repository root.
> 
> There is also an alternate \-z format recommended for machine parsing. In that format, the status field is the same, but some other things change. First, the '->' is omitted from rename entries and the field order is reversed (e.g 'from -> to' becomes 'to from'). Second, a NUL (ASCII 0) follows each filename, replacing space as a field separator and the terminating newline (but a space still separates the status field from the first filename). Third, filenames containing special characters are not specially formatted; no quoting or backslash-escaping is performed.

So yes, this looks interesting. However, I'm missing when both options were introduced. IIRC we now assume at least Git 1.8, although the SCM-API has no way to verify this (yet). If the `\-z` is supported for quite some time, it would probably be better to use it. The `GitStatusConsumer` has already quite some tests, so you should have good feedback on your changes. Looking forward on that patch.

**Comment by jira-importer** [comment]
**[Gary L Peskin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=garyp@firstech.com)** commented

I've attached a proposed patch file.  In getting further into this issue, I decided against using the "`-z`" option because it does not have each file on a separate line and switching to using it would cause too big of a change in the flow that drives the `GitStatusConsumer.consumeLine` method which is oriented towards a _line_ of output.

Instead, I detect when the "`--porcelain`" output is surrounded by quotation marks and I dequote it in a manner consistent with the way that Git quoted it.

The other main change was to get away from treating the file name as a URI since it is not a URI.  A filename can contain characters that are not allowed in a URI and trying to treat a filename as a relative URI is not a good fit.

Please review this and let me know what you think and if additional revisions are required.

Thanks.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

If I recall correctly, URI gave us an easy way to calculate the relative path without having OS issues. Be aware that when using File on Windows the pathseparator is \. It's a pretty big patch, so it'll take some time to verify it.

**Comment by jira-importer** [comment]
**[Gary L Peskin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=garyp@firstech.com)** commented

I think you'll find that most of the patch is just enhancing the testing components.  I agree that the URI does give an easy way to calculate the relative path.  However, using the URI approach failed in my use case due to the special characters.  I made sure that the code still passed all of the tests on both Windows 7 and Amazon Linux.  It would be great if you could find places where the new code fails and then I can add those to the test and enhance the code to take that into account.  There is now a Path.relativize() method that calculates the relative path but the Path class didn't come in until Java SE 1.7 so I didn't use it.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [68eecff6c782e18eb59840205c55d3d1631aa256](http://git-wip-us.apache.org/repos/asf/maven-scm/commit/68eecff6)
I've used a lot of the original patch, but was able to keep the URI as argument.
Thanks for the patch!
